src/Features/CSharp/Portable/UseIndexOrRangeOperator/CSharpUseIndexOperatorDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,224 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Threading;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.CodeAnalysis.CSharp.UseIndexOrRangeOperator
+{
+    using static Helpers;
+
+    /// <summary>
+    /// Analyzer that looks for code like: 
+    /// 
+    /// 1) `s[s.Length - n]` and offers to change that to `s[^n]`. and.
+    /// 2) `s.Get(s.Length - n)` and offers to change that to `s.Get(^n)`
+    ///
+    /// In order to do convert between indexers, the type must look 'indexable'.  Meaning, it must
+    /// have an int-returning property called 'Length' or 'Count', and it must have both an
+    /// int-indexer, and a System.Index-indexer.  In order to convert between methods, the type
+    /// must have identical overloads except that one takes an int, and the other a System.Index.
+    ///
+    /// It is assumed that if the type follows this shape that it is well behaved and that this
+    /// transformation will preserve semantics.  If this assumption is not good in practice, we
+    /// could always limit the feature to only work on a whitelist of known safe types.
+    /// 
+    /// Note that this feature only works if the code literally has `expr1.Length - expr2`.  If
+    /// code has this, and is calling into a method that takes either an int or a System.Index,
+    /// it feels very safe to assume this is well behaved and switching to `^expr2` is going to
+    /// preserve semantics.
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp), Shared]
+    internal partial class CSharpUseIndexOperatorDiagnosticAnalyzer : AbstractCodeStyleDiagnosticAnalyzer
```

Comment:
>AbstractCodeStyleDiagnosticAnalyzer [](start = 70, length = 35)

Should this inherit from [AbstractBuiltInCodeStyleDiagnosticAnalyzer](http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/CodeStyle/AbstractBuiltInCodeStyleDiagnosticAnalyzer.cs,7) that Sam recently introduced?

Reply:
Yes it should.  Thanks for catching that.  Still too used to my old patterns of making these things :)  I forgot about this new guy.